publish.go

```diff
@@ -0,0 +1,314 @@
+
+	var cleanTrack bool
+	if len(release.VersionCodes) != len(newVersionCodes) {
+		log.Warnf("Mismatching app count, removing (%v) versions from release: %s", release.VersionCodes, release.Name)
+		cleanTrack = true
+	} else {
+		log.Debugf("The number of App version codes (current and new) are equal")
+		release.VersionCodes = sortAndFilterVersionCodes(release.VersionCodes, newVersionCodes)
+	}
+
+	if cleanTrack {
+		log.Infof("Clearing version codes for release %v, the new versions will be %v", release.Name, newVersionCodes)
+		release.VersionCodes = newVersionCodes
+	}
+}
+
+// sortAndFilterVersionCodes sorts and filters two set of version codes, returns the higher for each.
+func sortAndFilterVersionCodes(currentVersionCodes googleapi.Int64s, newVersionCodes googleapi.Int64s) googleapi.Int64s {
+	sort.Slice(currentVersionCodes, func(a, b int) bool { return currentVersionCodes[a] < currentVersionCodes[b] })
+	sort.Slice(newVersionCodes, func(a, b int) bool { return newVersionCodes[a] < newVersionCodes[b] })
+
+	var filteredVersionCodes googleapi.Int64s
+	for i := 0; i < len(newVersionCodes); i++ {
+		log.Debugf("Searching for shadowing versions, comparing (%v) and (%v)", currentVersionCodes[i], newVersionCodes[i])
+		if currentVersionCodes[i] < newVersionCodes[i] {
+			log.Infof("Shadowing app found, removing current (%v) version, adding new (%v)", currentVersionCodes[i], newVersionCodes[i])
+			filteredVersionCodes = append(filteredVersionCodes, newVersionCodes[i])
+		} else {
+			log.Infof("Currently released app with (%v) version is higher than new (%v), app with new version code ignored", currentVersionCodes[i], newVersionCodes[i])
+			filteredVersionCodes = append(filteredVersionCodes, currentVersionCodes[i])
+		}
+	}
+	return filteredVersionCodes
+}
+
+// GetReleaseStatusFromConfig gets the release status from the config.
+func GetReleaseStatusFromConfig(configs Configs) string {
+	if configs.UserFraction != 0 {
+		log.Infof("Release is a staged rollout, %v of users will receive it.", configs.UserFraction)
+		return releaseStatusInProgress
+	}
+	return releaseStatusCompleted
+}
+
+// GetTrack gets the given track from the list of tracks of a given app.
+func GetTrack(configs Configs, allTracks []*androidpublisher.Track) (*androidpublisher.Track, error) {
+	currentTrack := configs.Track
+	for _, track := range allTracks {
+		if currentTrack == track.Track {
+			log.Debugf("Current track found, name '%s'", currentTrack)
+			return track, nil
+		}
+	}
+
+	return &androidpublisher.Track{
+		Releases:        []*androidpublisher.TrackRelease{},
+		Track:           currentTrack,
+		ServerResponse:  googleapi.ServerResponse{},
+		ForceSendFields: []string{},
+		NullFields:      []string{},
+	}, nil
+}
+
+// GetAllTracks lists all tracks for a given app.
+func GetAllTracks(configs Configs, service *androidpublisher.Service, appEdit *androidpublisher.AppEdit) ([]*androidpublisher.Track, error) {
```

Comment:
Please pass in packageName only, instead of the Configs object. It is the only field used, and in my opnion is bad practice to pass the catch-all config object. If multiple similar stuff is needed, please define a struct that contains neccessary fields.

Reply:
Updated args.